rfd/0201-native-auto-approval.md

```diff
@@ -0,0 +1,318 @@
+enforce a stricter match condition for automatic approval rules compared to
+notification rules.
+
+The submitted form will be converted into an access monitoring rule that would
+look like the following yaml. More details about the access monitoring rule
+changes will be provided in following sections.
+
+The `Role Name` input is converted into a `contains_all` expression. In this
+example, we have the roles "cloud-dev" and "cloud-stage". The set of requested
+roles that can match this condition are either `set("cloud-dev")`, `set("cloud-stage")`,
+or `set("cloud-dev", "cloud-stage")`.
+
+Note that the converted conditions for requester traits use AND logic across traits, and OR
+logic within a trait. The user must match at least one of the values among each
+configured trait. For example, given the following converted AMR, a user is
+pre-approved if they are any level "L1" or "L2" and they are on the "Cloud" team
+and they are located in "Seattle". If the user is level "L1" and located in
+"Seattle", but they are on the "Tools" team, they would not be pre-approved. If
+the user requires more control over the condition matching, they will need to
+edit the access monitoring rule yaml directly.
+
+```yaml
+kind: access_monitoring_rule
+version: v1
+metadata:
+  name: dev-pre-approved
+spec:
+  subjects:
+    - access_request
+  states:
+    - approved
+  condition: >
+    contains_all(set("cloud-dev", "cloud-stage"), access_request.spec.roles) &&
+    contains_any(user.traits["level"], set("L1", "L2")) &&
+    contains_any(user.traits["team"], set("Cloud")) &&
+    contains_any(user.traits["location"], set("Seattle"))
+  automatic_approval:
+    name: teleport
+  notification:
+    name: slack
+    recipients: ["#dev-cloud"]
+```
+
+The `Access Monitoring Rules` overview page will be modified to display both
+notification rules, as well as automatic approval rules. This page will allow
+user to see a quick overview of the automatic approvals currently enabled. The
+overview simply displays the access monitoring rule name, plugin/integration
+name, and the roles that are automatically approved. Users will need to click
+on the **View** button to see the actual conditions for auto approval.
+
+![view-amr](assets/0201-view-amr.png)
+
+## Details
+This feature will be supported by a new internal Teleport service. This
+automatic approval service will function similarly to existing access plugins.
+It will be running as part of the Teleport Auth Service by default.
+
+The automatic approval service relies on a similar workflow that supports access
+request notification routing with access monitoring rules. The service watches
+for Access Monitoring (AMR) events and Access Request (AR) events. If an
+incoming AR matches an existing AMR condition, then the service will attempt to
+automatically approve the request.
+
+### Access Monitoring Rule
+The AMR now supports a configurable `automatic_approval.name` spec. This field
```

Comment:
Based on our UX discussion a couple of weeks ago, I'm thinking we could actually treat the absence of the plugin name as the "builtin approval". It's a minor implementation detail but would be consistent with how we're thinking of presenting this in the UI.